File: cpp/3304_kth_character/kth_character.cpp
#include <cassert>
#include <tuple>
#include <vector>

class Solution {
  public:
    char kthCharacter(int k) {
      int ans = 0;
      int t;
      while (k != 1) {
        t = std::__lg(k);
        if ((1 << t) == k) {
          t--;
        }
        k = k - (1 << t);
        ans++;
      }
      return 'a' + ans;
    }
};

int main() {
  std::vector<std::tuple<int, char>> tests {
    {5, 'b'},
    {10, 'c'},
  };

  for (auto &[k ,ans] : tests) {
    assert(Solution().kthCharacter(k) == ans);
  }
}

File: golang/3304_kth_character/kth_character.go
// Package main ...
package main

import (
	"math/bits"
	"testing"

	"github.com/stretchr/testify/assert"
)

func kthCharacter(k int) byte {
	ans := 0
	for k != 1 {
		t := bits.Len(uint(k)) - 1
		if 1<<t == k {
			t--
		}
		k -= 1 << t
		ans++
	}
	return byte('a' + ans)
}

func main() {
	tests := []struct {
		k   int
		ans byte
	}{
		{5, 'b'},
		{10, 'c'},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, test.ans, kthCharacter(test.k), index)
	}
}
